File: backend/app/utils/pagination.py
def build_meta(page: int, page_size: int, total: int) -> dict:
    return {"page": page, "pageSize": page_size, "total": total}


def paginate_items(items, page: int, page_size: int):
    total = len(items)
    start = (page - 1) * page_size
    end = start + page_size
    return items[start:end], build_meta(page, page_size, total)

File: frontend/src/components/Pagination.js
export default function Pagination({ meta, onPageChange }) {
  if (!meta || !meta.total || meta.total <= meta.pageSize) {
    return null;
  }

  const totalPages = Math.max(1, Math.ceil(meta.total / meta.pageSize));

  return (
    <div className="pagination">
      <button
        className="btn btn-light"
        type="button"
        onClick={() => onPageChange(meta.page - 1)}
        disabled={meta.page <= 1}
      >
        Poprzednia
      </button>

      <span className="pagination-info">
        Strona <strong>{meta.page}</strong> z <strong>{totalPages}</strong>
      </span>

      <button
        className="btn btn-light"
        type="button"
        onClick={() => onPageChange(meta.page + 1)}
        disabled={meta.page >= totalPages}
      >
        Następna
      </button>
    </div>
  );
}